[WIP] Complete last stage of the project

denshykk opened the pull request:
**Description**
Running a cinema theatre is no easy business. To help our friends, let's add statistics to your program. The stats will show the current income, total income, the number of available seats, and the percentage of occupancy.

In addition, our friends asked you to take care of a small inconvenience: it's not good when a user can buy a ticket that has already been purchased by another user. Let's fix this!

**Objectives**
Now your menu should look like this:

```
1. Show the seats
2. Buy a ticket
3. Statistics
0. Exit
```
When the item Statistics is chosen, your program should print the following information:

- The number of purchased tickets;
- The number of purchased tickets represented as a percentage. Percentages should be rounded to 2 decimal places;
- Current income;
- The total income that shows how much money the theatre will get if all the tickets are sold.

The rest of the menu items should work the same way as before, except the item Buy a ticket shouldn't allow a user to buy a ticket that has already been purchased.

If a user chooses an already taken seat, print That ticket has already been purchased! and ask them to enter different seat coordinates until they pick an available seat. Of course, you shouldn't allow coordinates that are out of bounds. If this happens, print Wrong input! and ask to enter different seat coordinates until the user picks an available seat.

**Examples**
The greater-than symbol followed by a space (> ) represents the user input. Note that it's not part of the input.

```
Enter the number of rows:
> 6
Enter the number of seats in each row:
> 6

1. Show the seats
2. Buy a ticket
3. Statistics
0. Exit
> 3

Number of purchased tickets: 0
Percentage: 0.00%
Current income: $0
Total income: $360

1. Show the seats
2. Buy a ticket
3. Statistics
0. Exit
> 2

Enter a row number:
> 1
Enter a seat number in that row:
> 1

Ticket price: $10

1. Show the seats
2. Buy a ticket
3. Statistics
0. Exit
> 3

Number of purchased tickets: 1
Percentage: 2.78%
Current income: $10
Total income: $360

1. Show the seats
2. Buy a ticket
3. Statistics
0. Exit
> 2

Enter a row number:
> 1
Enter a seat number in that row:
> 1

That ticket has already been purchased!

Enter a row number:
> 10
Enter a seat number in that row:
> 20

Wrong input!

Enter a row number:
> 4
Enter a seat number in that row:
> 4

Ticket price: $10

1. Show the seats
2. Buy a ticket
3. Statistics
0. Exit
> 1

Cinema:
  1 2 3 4 5 6
1 B S S S S S
2 S S S S S S
3 S S S S S S
4 S S S B S S
5 S S S S S S
6 S S S S S S

1. Show the seats
2. Buy a ticket
3. Statistics
0. Exit
> 3

Number of purchased tickets: 2
Percentage: 5.56%
Current income: $20
Total income: $360

1. Show the seats
2. Buy a ticket
3. Statistics
0. Exit
> 0
```

denshykk committed:
Complete last stage of the project

olegvovk96 commented on src/service/Cinema.java:
```diff
@@ -0,0 +1,58 @@
+package service;
+
+public class Cinema {
```
This class is placed in the same package as CinemaService, you can use the default access modifier.

denshykk replied:
Changed

olegvovk96 commented on src/service/Cinema.java:
```diff
@@ -0,0 +1,58 @@
+    static class Seat {
+
+        int row;
+        int columns;
```
Typo here. It should be singular noun.

denshykk replied:
Changed to singular noun instead of plural

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    private Cinema initializeCinema() {
+        Scanner scanner = new Scanner(System.in);
```
Can you create a single instance of Scanner for CinemaService?

denshykk replied:
Extracted Scanner initialization and creation from methods to field/constructor

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+
+        for (int i = 0; i < cinema.getRows(); i++) {
+            String[] row = new String[cinema.getColumns()];
+            Arrays.fill(row, "S");
```
Can you make "S" and "B" constants?

denshykk replied:
Fixed

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+
+            switch (input) {
+                case "0" -> {
+                    return;
```
It would be better to close Scanner at the end of the program explicitly.

denshykk replied:
fixed

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+
+    private void showSeats() {
+        System.out.print("Cinema:\n ");
+        for (int i = 1; i <= cinema.getSeats()[0].length; i++) {
+            System.out.print(" " + i);
+        }
```
Can you use Stream here?

denshykk replied:
Fixed

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+        for (String[] seat : cinema.getSeats()) {
+            System.out.print(counter);
+            for (String s : seat) {
+                System.out.print(" " + s);
```
Can you avoid using string concatenation? It uses implicitly StringBuilder under the hood.

denshykk replied:
fixed

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+
+    private void buyTicket() {
+        try {
+            Cinema.Seat seat = seatValidation();
```
seatValidation() - this method does more than it should be.

denshykk replied:
divided seatValidation() method's obligation.

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+                Percentage: %s%%
+                Current income: $%d
+                Total income: $%d
+                                      
```
Can you remove this empty row?

denshykk replied:
fixed

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+
+    private int getPrice(Cinema.Seat seat) {
+        int price;
+        boolean isFrontSide = seat.row <= cinema.getRows() / 2;
```
Can you use the access method to access `seat.row`?

denshykk replied:
Created getters and used them instead of direct field invocation

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    private void showSeats() {
+        System.out.print("Cinema:\n ");
+        for (int i = 1; i <= cinema.getSeats()[0].length; i++) {
+            System.out.print(" " + i);
+        }
+
+        System.out.println();
+
+        int counter = 1;
+        for (String[] seat : cinema.getSeats()) {
+            System.out.print(counter);
+            for (String s : seat) {
+                System.out.print(" " + s);
+            }
+            System.out.println();
+            counter++;
+        }
```
Can you add tabulation when printing the matrix?

![image](https://user-images.githubusercontent.com/27965136/145179726-2e3ca55b-5b52-4e80-8de8-4bdca3d6cb4f.png)

denshykk replied:
Fixed.

denshykk committed:
Fixes according to PR reviews

- Change access modifiers
- Correct typo, empty lines
- Extract Scanner from methods, make it a field and close it
- Make "S" and "B" local constants
- Replace standard for loop with Stream
- Divide seatValidation method's obligations between two methods
- Use accessors instead of field invocation
- Add tabulation when printing the matrix

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    private void createSeats() {
+        final String S = "S";
```
That's not a constant. Can you please use a more meaningful name than **S**?

denshykk replied:
Fixed

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+    private void showSeats() {
+        System.out.print("Cinema:\n ");
+        IntStream.rangeClosed(1, cinema.getSeats()[0].length)
+                .mapToObj(i -> " \t" + i)
```
The tabulation is related to IO operation, so you can just do it in the forEach operation with a lambda expression.

denshykk replied:
Shorten stream

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+        for (String[] seat : cinema.getSeats()) {
+            System.out.print(counter);
+            for (String s : seat) {
+                System.out.print("\t".concat(s));
```
Can you use `printf` with the parameter?

denshykk replied:
Solved

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    private void seatValidation(Cinema.Seat seat) throws NonExistingSeatException, AlreadyTakenSeatException {
+        final String B = "B";
```
That's not a constant. Can you please use a more meaningful name than **B**?

denshykk replied:
Fixed

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,179 @@
+    private void seatValidation(Cinema.Seat seat) throws NonExistingSeatException, AlreadyTakenSeatException {
+        final String B = "B";
+
+        if (seat.row() < 1 || seat.row() > cinema.getRows() || seat.column() < 1 || seat.column() > cinema.getColumns()) {
```
This condition is pretty long and hard to read. 
Can you move column conditions to the new line?

denshykk replied:
Fixed

denshykk committed:
Code reformat

- Make "B" constant and name it soldSeat
- Make "S" constant and name it availableSeat
- Shorten Stream
- Make printf instead of concat
- Reformat code for better readability

denshykk committed:
Replace print and string concatenation with printf

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,178 @@
+
+public class CinemaService {
+
+    private static final String availableSeat = "S";
+    private static final String soldSeat = "B";
```
The constant naming convention is uppercase with words separated by underscores.
e.g. 
```
private static final String AVAILABLE_SEAT = "S";
```
Can you follow this standard?

denshykk replied:
FIxed

olegvovk96 commented on src/service/CinemaService.java:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    private void showSeats() {
+        System.out.print("Cinema:\n ");
+        IntStream.rangeClosed(1, cinema.getSeats()[0].length)
+                .forEach(i -> System.out.printf(" \t%d", i));
```
Can you remove spaces after `Cinema:\n` and before `\t%d`?

denshykk replied:
Fixed

denshykk committed:
Correct constant naming convention; Remove redundant spaces

olegvovk96 commented:
Nice work!
